### ISSUE TITLE
Is it possible to get a complete go implementation by using golang call rust?

### ISSUE THREAD

**Issue by shootingstar123** [BLAKE3-team/BLAKE3#273]
I know there is an incomplete implementation of blakes by golang, well it doesn't support mutli-thread, nor full interfaces. I' am wondering if I can implement blake3 of golang by using golang to call rust, with FFI or CGO. I am going to use BLAKE3 in a golang project receiving file stream. I am not familiar with rust, Is there any difficulty to do that?

**Comment by jedisct1** [comment]
There are excellent, pure Go implementations: 

- https://github.com/lukechampine/blake3
- https://github.com/zeebo/blake3

**Comment by oconnor663** [comment]
Agreed with Frank, both of those implementations are mature and full-featured, and I'd encourage any Go application to use them. That said, as far as I know, they don't support multithreaded/multi-goroutine hashing. (Neither does the official C implementation, for what it's worth.) If that was a requirement for you, and you didn't want to just shell out to `b3sum`, it is possible to call the Rust implementation from Go. This requires both unsafe Go and unsafe Rust, and I wouldn't suggest doing it in important code unless you were comfortable with both languages.

Here's a demo of how you stitch it together: https://github.com/oconnor663/go_calling_rust_example. The high level steps are:
- expose one or more `extern "C"` functions from Rust that wrap the parts of the `blake3` crate you want to use
- add declarations for those functions in a cgo block
- configure the Rust code to build a staticlib
- use `#cgo LDFLAGS` to make Go link against that staticlib

**Comment by shootingstar123** [comment]
> 

Thanks frank! I know one of the go implemention like zeebo.Well both implementions can't support multithreaded. But I need to use blake3 to process relatively big data file and I have enough cpu cores in my server. I will make a test for the performance difference between rust multi-threaded and the two go implementions.

**Comment by shootingstar123** [comment]
> Agreed with Frank, both of those implementations are mature and full-featured, and I'd encourage any Go application to use them. That said, as far as I know, they don't support multithreaded/multi-goroutine hashing. (Neither does our official C implementation, for what it's worth.) If that's a requirement for you, and you can't just shell out to `b3sum`, it is possible to call the Rust implementation from Go. This requires both unsafe Go and unsafe Rust, and I wouldn't suggest doing it in important code unless you were comfortable with both languages.
> 
> Here's a demo of how you stitch it together: https://github.com/oconnor663/go_calling_rust_example. The high level steps are:
> 
> * expose one or more `extern "C"` functions from Rust that wrap the parts of the `blake3` crate you want to use
> * add C declarations for those functions in a cgo block
> * configure the Rust code to build a staticlib
> * use `#cgo LDFLAGS` to make Go link against that staticlib
> 
> So yes, this is quite involved, and I'm not entirely sure I've even done it right. Please be careful with it :)

Thank you Jack for your patient detail reply and calling demo. I just tried but met some errors, "undefined reference to `dlsym'", I need some time to trick it. I also prefer to use pure go implemention after got your advices. So firstly I will make a test for the performance difference between go single-threaded and rust multi-threaded implementions for big files. If not significant difference, I will use go implemention

**Comment by oconnor663** [comment]
Note that to get the benefit of multithreaded hashing, you usually need to `mmap` the files your reading, which adds some more complexity and unsafety. One advantage of shelling out to `b3sum` is that it takes care of those details for you, and it doesn't require any tricky FFI hacks. If your files are big enough, the process spawning overhead might be negligible. That could be worth benchmarking.
